### ISSUE TITLE
Quick input: consider to better indicate list focus

### ISSUE THREAD

**Issue by Tyriar** [microsoft/vscode#93916]
https://github.com/microsoft/vscode/issues/93641

Version: 1.44.0-insider (user setup)
Commit: d7d1147d6b764a48b789000c03088ee9afabcd8a
Date: 2020-03-31T08:23:56.446Z
Electron: 7.1.11
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Windows_NT x64 10.0.18362

---

Using NVDA:

1. ctrl+p
2. 'term ' (announced: 2 results)
3. Press down (announced: term list, 1.1 node, ...)
4. **Press tab (silence)**
5. Press tab, the icon is selected (toolbar, rename terminal button, ...)

**Comment by Tyriar** [comment]
The same happens with other quick picks too like go to file.

**Comment by isidorn** [comment]
I can reproduce this, and we are doing everything correct here imho just NVDA decides not to read anything. VoiceOver behaves correctly.
Idealy the first tab would focus imedaly on the icon, not sure if that is possible though.

**Comment by Tyriar** [comment]
@isidorn what does voiceover do at step 4?

**Comment by chrmarti** [comment]
The first tab focuses on the list. My understanding is that that is how it should be from an accessibility point of view. @isidorn ?

**Comment by bpasero** [comment]
I am not seeing an immediate issue here to fix and this is pending answer from a18y experts, so moving to April.

**Comment by Tyriar** [comment]
Someone else created an issue about this the same day but from the angle that it doesn't show a focus indicator https://github.com/microsoft/vscode/issues/93907, we should either remove the ability to tab to the list or if that's not good from a a11y perspective we should add a special focus indicator to it (an outline around the selected item?).

**Comment by moshfeu** [comment]
I'm voting for outline ☝️

**Comment by chrmarti** [comment]
The list does shows an outline when it has focus and there is no item focused. We usually focus an item. That's the List's built-in behavior.

**Comment by Tyriar** [comment]
@chrmarti there is no visual difference between implied focus because the textbox is focused:

![image](https://user-images.githubusercontent.com/2193314/78672224-37cf1580-7895-11ea-9047-064681d531cb.png)

and actual focus when you tab:

![image](https://user-images.githubusercontent.com/2193314/78672253-41f11400-7895-11ea-97a6-62769800a44d.png)

I think the second one should look like this or something similar:

![image](https://user-images.githubusercontent.com/2193314/78672347-5e8d4c00-7895-11ea-944e-b1b3ed8e6ff4.png)

**Comment by chrmarti** [comment]
That's on purpose, by default the implied focus looks different, but because you can navigate the list with up-/down-arrows, the implied focus was changed to look the same as the focus.

Not sure what the original discussion around this was, but this makes sense to me.

**Comment by Tyriar** [comment]
@chrmarti to me it feels broken when you transition to this:

![image](https://user-images.githubusercontent.com/2193314/78676002-b4182780-789a-11ea-8b19-55418e2acc53.png)

Because you lose the "real" focus indicator, combine that with NVDA not announcing anything and it felt very broken.

**Comment by chrmarti** [comment]
The screenreader should already have read out the entry at that point, so it does not repeat it, I assume.

We are using `aria-activedescendant` to get the entry read out while the input box has focus, not sure if and how we could get screenreaders to read the entry again when the focus moves to the list.

**Comment by bpasero** [comment]
I find the outline around the list item a bit weird because we do not have this kind of styling in any other list or tree widget.  We typically distinguish focus via the selection color:

![image](https://user-images.githubusercontent.com/900690/79536010-47addd00-807f-11ea-9bda-1a03e9789329.png)

vs

![image](https://user-images.githubusercontent.com/900690/79536026-4ed4eb00-807f-11ea-9ba8-a86cf46bcbc8.png)

If we have to introduce an outline I would probably rather expect that to be around the entire list widget, same as we have it when no item is selected on lists:

![image](https://user-images.githubusercontent.com/900690/79536076-6ca25000-807f-11ea-9cf9-50c3ab365034.png)

**Comment by Tyriar** [comment]
@bpasero well the problem is we use implicit/linked focus in the quick pick but style it the same as regular focus. I still think the outline would be good because it's clear what is actually focused and this is mainly only exposed for accessibility reasons, we have also established outline to mean focus (in addition to solid bg).

Outline around textbox = textbox is focused:

![image](https://user-images.githubusercontent.com/2193314/79571362-ee8a7d00-806f-11ea-8ed6-0f574de1677b.png)

Outline around close button = close button is focused:

![image](https://user-images.githubusercontent.com/2193314/79571428-09f58800-8070-11ea-845f-a40ea31e19b6.png)

No outline = textbox is no longer focused 🤷 

![image](https://user-images.githubusercontent.com/2193314/79571487-1da0ee80-8070-11ea-9091-a1a30c7ed22a.png)

**Comment by Tyriar** [comment]
cc @misolori

**Comment by misolori** [comment]
Related to this is how our list focus/active selection can be confusing:

![image](https://user-images.githubusercontent.com/35271042/79770090-cc983100-82e1-11ea-9937-396e17770186.png)

![image](https://user-images.githubusercontent.com/35271042/79779898-f9534500-82ef-11ea-8523-31d0faa3ab18.png)

![image](https://user-images.githubusercontent.com/35271042/79779935-06703400-82f0-11ea-8e57-ce6e251ff363.png)

![image](https://user-images.githubusercontent.com/35271042/79779947-0b34e800-82f0-11ea-9f99-4959067b366d.png)

**Comment by Tyriar** [comment]
What if multi-select leveraged the same border style I'm proposed, making it a more common paradigm to have a "higher priority" focus indicator to clear up ambiguities like this.

Shift+click:

![image](https://user-images.githubusercontent.com/2193314/79784379-ff005900-82f6-11ea-94d0-de1da176037f.png)

Ctrl+click:

![image](https://user-images.githubusercontent.com/2193314/79784320-eb54f280-82f6-11ea-894c-78cff82cd9f1.png)

**Comment by bpasero** [comment]
That would be a model as Windows has it imho:

![image](https://user-images.githubusercontent.com/900690/79829986-bd2dec00-83a4-11ea-96ea-a51e972d5570.png)

I somewhat like it but would need to try it out to provide more feedback. This would requires changes to the tree/list rendering. 

//cc @joaomoreno

**Comment by joaomoreno** [comment]
I like it!

**Comment by Tyriar** [comment]
Move explorer request to https://github.com/microsoft/vscode/issues/95768

**Comment by bpasero** [comment]
@Tyriar this is not explorer specific? this applies to every list and tree?

**Comment by Tyriar** [comment]
@bpasero I think the implicit focus when textbox is focused is a quick pick thing?

**Comment by bpasero** [comment]
@Tyriar yeah, I was referring to https://github.com/microsoft/vscode/issues/95768 but I think Joao already made clear that such a change impacts every tree/list usage outside of maybe quick pick. 

Still, for quick pick I would want to wait for the result of https://github.com/microsoft/vscode/issues/95768 to see if we can benefit from the focus ring also for quick pick. Maybe we could still align the visuals for quick pick and other lists/trees.
